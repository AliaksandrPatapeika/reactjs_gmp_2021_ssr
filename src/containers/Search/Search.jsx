import './Search.less';

import {assign} from 'lodash';
import {useRouter} from 'next/router';
import React, {
  useCallback, useEffect, useRef, useState
} from 'react';
import {useDispatch, useSelector} from 'react-redux';

import Button from '../../components/Button';
import {getQueryString} from '../../utils';
import SearchInput from '../SearchInput';

const Search = () => {
  const query = useSelector((state) => state.query);
  const dispatch = useDispatch();
  const router = useRouter();
  const searchButtonRef = useRef();
  const {search} = router.query;
  const [value, setValue] = useState('');

  const onClickHandler = () => {
    router.push({
      pathname: '/search',
      search: getQueryString(assign(query, {search: value}))
    });
    // , '/search page'); // (router.push 2nd parameter) Implement route masking
  };
  const handleKeyDown = useCallback(
    (event) => {
      if (event.key === 'Enter') {
        searchButtonRef.current.click();
      }
    },
    []
  );

  useEffect(() => {
    if (search) {
      setValue(search);
    }
    window.addEventListener('keydown', handleKeyDown);
    return () => {
      window.removeEventListener('keydown', handleKeyDown);
    };
  }, [dispatch, handleKeyDown, search]);
  return (
    <div className="searchContainer">
      <SearchInput
        value={value}
        onChange={(event) => setValue(event.target.value)}
      />
      <Button
        className="searchButton"
        ref={searchButtonRef}
        title="SEARCH"
        onClick={onClickHandler}
      />
    </div>
  );
};

export default Search;
