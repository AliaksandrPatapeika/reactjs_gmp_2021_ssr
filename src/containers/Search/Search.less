@import "../../less/variables";

.searchContainer {
	display: flex;
	margin-top: 33px;

	.searchButton {
		cursor: pointer;
		padding: 16px 77px;
		border-radius: 4px;
		color: @white;
		font-size: 22px;
		font-family: @NetflixSans-Regular;
		background-color: @search-red;

		&:hover,
		&:focus {
			background-color: @red-hover;
		}
	}
}
