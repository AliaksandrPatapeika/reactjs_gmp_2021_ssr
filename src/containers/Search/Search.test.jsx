import React from 'react';
import {Provider} from 'react-redux';
import renderer from 'react-test-renderer';
import configureMockStore from 'redux-mock-store';

import Search from './Search';

jest.mock('next/router', () => ({
  useRouter: () => ({
    query: {search: 'movie'}
  })
}));

describe('Search', () => {
  test('renders correctly', () => {
    const mockStore = configureMockStore();
    const initialState = {
      query: {
        filter: '',
        sortBy: 'release_date',
        sortOrder: 'desc'
      }
    };
    const store = mockStore(initialState);
    const search = renderer
      .create(
        <Provider store={store}>
          <Search />
        </Provider>
      )
      .toJSON();

    expect(search)
      .toMatchSnapshot();
  });
});
