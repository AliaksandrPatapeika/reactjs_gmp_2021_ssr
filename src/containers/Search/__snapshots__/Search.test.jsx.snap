// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search renders correctly 1`] = `
<div
  className="searchContainer"
>
  <input
    autoComplete="off"
    className="searchInput"
    onChange={[Function]}
    placeholder="What do you want to watch?"
    value=""
  />
  <button
    className="searchButton"
    disabled={false}
    id={null}
    onClick={[Function]}
    type="button"
  >
    SEARCH
    
  </button>
</div>
`;
